feat: Support Persistence features

jimtng opened the pull request:
Resolve #50 
Add Persistence methods to Items. It accepts a relative duration in addition to ZonedDateTime, e.g. 
```
Item1.average_since 12.hours #or
Item1.average_since ZonedDateTime.now.minusHours(12)
```

To specify the persistence service: (I dropped the idea of named parameter)
```
Item1.last_update 'influxdb'
Item1.sum_since 1.hour, :influxdb #it could accept a symbol
```

jimtng commented:
@boc-tothefuture if you're happy with this implementation, would you mind writing the tests? It would need to install a persistence extension (either rrd4j, or influxdb?), setting it as default, generating some data, etc.

boc-tothefuture commented:
I will take a look at how to test this. 

What happens if you don't specify a persistence service with the calls?

jimtng commented:
> What happens if you don't specify a persistence service with the calls?

It will use the system default persistence service. It is specified in the gui, or in config file although I'm not sure how exactly.

boc-tothefuture commented:
Is there any value, in addition to the work you did, add a persistence block to group multiple persistence operations together under a single service?  Much like we did with units?

So, just an example.
```ruby
persistence(:influxdb) do 
  Item1.average_since
  Item2.sum_since
  etc...
end
```

We would keep both styles.

boc-tothefuture commented:
FYI - build is failing is because a method was added without documentation.

jimtng commented:
I've added a yard documentation for the `previous_state` method. How do we document the other ones that are defined with define_method ?

boc-tothefuture commented:
How do you feel about the addition of the block method?

jimtng commented:
> Is there any value, in addition to the work you did, add a persistence block to group multiple persistence operations together under a single service? Much like we did with units?
> 
> So, just an example.
> 
> ```ruby
> persistence(:influxdb) do 
>   Item1.average_since
>   Item2.sum_since
>   etc...
> end
> ```
> 
> We would keep both styles.

I didn't see this message. This looks good. I wasn't aware of such usage with units until you've pointed it out here.

I'll have a play with this and update this PR.

jimtng commented:
@boc-tothefuture Please review the persistence block that I've just added. I've briefly tested it.

jimtng commented:
Note: Once #122 is merged, we'll need to remove this piece of code from persistence.rb since it is no longer relevant.

```
                if timestamp.respond_to? 'to_ms'
                  timestamp = ZonedDateTime.now.plus(Java::JavaTime::Duration.ofMillis(timestamp.to_ms))
```

jimtng commented:
I've removed the `to_ms` reference and also modified it to accept `TemporalAmount`, which is the parent interface of `Duration`, so technically it can also accept `java.time.Period`.

jimtng commented:
Please don't merge yet. I think I can write the tests too, just have been a bit busy with life + also working on another PR for doing comparisons.

jimtng commented:
@boc-tothefuture it's ready for review now

boc-tothefuture commented:
Awesome!  Thanks.  Lets merge in #130 first.

boc-tothefuture commented:
Do we think there is no need to test with rrd or something similar?

jimtng commented:
This is read to merge

> Do we think there is no need to test with rrd or something similar?

Currently it's tested by installing rrd4j and using it using the persistence block. It doesn't really verify the values. It just assumes that if the calls were successful, it will continue to the end, i.e. no errors.

boc-tothefuture commented:
I totally missed that it installed that the rrd4j.
